### PR TITLE
Add armv8r-none-eabihf target for the Cortex-R52.

### PR DESCRIPTION
**Pull request by chrisnc**

### CODE REVIEW

**Comment by WaffleLapkin** [comment]
What is the context there? Does the new target agree with the target policy? what is the reason for removing `-unknown` parts?

**Comment by chrisnc** [comment]
Yes, as @moridinga says, `unknown` is unnecessary here, and the other `{arm,thumb}*-none-eabi*` targets don't include it, so I removed it from the v7r targets for consistency.

As for the armv8r addition, I was unsure how this should be handled by the target policy. It's not much different from any of the existing arm/thumb-none-eabi targets, just setting the version and minimum features for this specific sub-family, using the same convention as thumbv7em-none-eabihf of picking the least-capable CPU that implements this version of the architecture and setting the default features to match. The R52 has been around several years now, but currently there's not a convenient way to target it with rust. armv7r-none-eabihf will work, but care must be taken because the set of floating-point features available differs substantially, and the defaults don't match any available R52 (and by default you won't get any v8 instructions of course).

Let me know which changes I should undo or any additional changes needed to align this PR with the target policy.

**Comment by chrisnc** [comment]
r? rust-lang/compiler

**Comment by chrisnc** [comment]
Bump?

File: src/doc/rustc/src/platform-support.md
Comment (nagisa) [inline src/doc/rustc/src/platform-support.md:149]: I think this target with its own caveats and all would benefit from a dedicated target page describing major points (such as around the floating point support.)

Yes, the `armv7r` targets don’t have one, but it wouldn’t be a terrible time to start – it’ll also be a great place to note down as to who the maintainer(s) of the target are.

File: compiler/rustc_target/src/spec/armv8r_none_eabihf.rs
Comment (nagisa) [inline compiler/rustc_target/src/spec/armv8r_none_eabihf.rs:20]: The wording in the referenced documentation suggests that this is a "NEON is optional” rather than “you get fp XOR neon” choice. This is quite standard in other ARM targets, unlike the XOR choice.

File: compiler/rustc_target/src/spec/armv8r_none_eabihf.rs
Comment (nagisa) [inline compiler/rustc_target/src/spec/armv8r_none_eabihf.rs:27]: I was confused about this, so I investigated. You get C compilers produce 1-byte large enums only with the `-fshort-enums` flag on armv8r targets, rather than by default, so this comment is not quite correct. It seems like it might have been added (to armv7r, which follows this behaviour) by mistake or propagated from some other targets.

**Review by nagisa — COMMENTED** [review]
> As for the armv8r addition, I think the operative clause of the target policy is this bullet:

Usually what we request target addition PRs to do is to copy over the ckecklist for the relevant tiers and confirm all the points are true, similar to what’s seen in [this comment](https://github.com/rust-lang/rust/pull/108795#issue-1610469403) most recently.

**Comment by nagisa** [comment]
(also, really really sorry for the insane delay!)

File: compiler/rustc_target/src/spec/armv8r_none_eabihf.rs
Comment (chrisnc) [reply compiler/rustc_target/src/spec/armv8r_none_eabihf.rs:20]: I'm not sure I follow, or if my comment wasn't clear. The critical part I was trying to capture here is that the non-neon version also has only 16 DP registers; there is no version that has 32 DP regs but no neon.

Comment (chrisnc) [reply compiler/rustc_target/src/spec/armv8r_none_eabihf.rs:20]: Also there is no `nofp` version of the R52, as I understand the documentation.

File: compiler/rustc_target/src/spec/armv8r_none_eabihf.rs
Comment (chrisnc) [reply compiler/rustc_target/src/spec/armv8r_none_eabihf.rs:27]: Hrm... I just experimented and stock `gcc` will give the size of an enum as one byte without using `-fshort-enums`:
```
$ arm-none-eabi-gcc --version
arm-none-eabi-gcc (Arm GNU Toolchain 12.2 (Build arm-12-mpacbti.34)) 12.2.1 20230214
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
$ cat foo.c
enum foo {
    FOO,
};

unsigned long f(void) {
    return sizeof(enum foo);
}
$ arm-none-eabi-gcc -march=armv8-r -c -o foo.o -Os foo.c
$ arm-none-eabi-objdump -d foo.o

foo.o:     file format elf32-littlearm


Disassembly of section .text:

00000000 <f>:
   0:	e3a00001 	mov	r0, #1
   4:	e12fff1e 	bx	lr
```

However, the same experiment with `clang` 16.0.6 and `-target armv8r-none-eabi` shows `4`...

File: src/doc/rustc/src/platform-support.md
Comment (chrisnc) [reply src/doc/rustc/src/platform-support.md:149]: Thanks, I'll take a stab at this.

File: compiler/rustc_target/src/spec/armv8r_none_eabihf.rs
Comment (chrisnc) [reply compiler/rustc_target/src/spec/armv8r_none_eabihf.rs:27]: Looks like AAPCS explicitly allows either convention: https://github.com/ARM-software/abi-aa/blob/main/aapcs32/aapcs32.rst#enumerated-types

Comment (chrisnc) [reply compiler/rustc_target/src/spec/armv8r_none_eabihf.rs:27]: Looks like all other arm*-none-eabi* targets in Rust are using the small enum ABI, though the discrepancy with recent `clang` is a little worrying. Will have to do some archaeology to understand if/when `clang` ever had this default, as the comments claim.

File: compiler/rustc_target/src/spec/armv8r_none_eabihf.rs
Comment (nagisa) [reply compiler/rustc_target/src/spec/armv8r_none_eabihf.rs:20]: Ah, so my interpretation of that comment was as such:

> supports either (fp-armv8 (which is single-precision only with 16 dp regs)) or (neon-fp-armv8 (which has do 32 double precision registers)).

hence the fp-armv8 XOR neon-fp-armv8 interpretation. Would it be better to copy over the note from the [upstream documentation](https://developer.arm.com/documentation/100026/0102/Advanced-SIMD-and-floating-point-support/About-the-Advanced-SIMD-and-floating-point-support) and then amend it with information about registers if you must? For example as such:

> The Cortex-R52 processor supports single-precision floating-point instructions (fp-armv8) and can optionally support the double-precision floating-point and Advanced SIMD instructions (neon-fp-armv8).
>
> In fp-armv8 only 16 double-precision registers are available. Available set of registers expand to 32 double-precision registers when neon-fp-armv8 is enabled.

or something.

File: compiler/rustc_target/src/spec/armv8r_none_eabihf.rs
Comment (nagisa) [reply compiler/rustc_target/src/spec/armv8r_none_eabihf.rs:27]: Yea, I only tested clang myself, so I sort-of assumed it might be similar with GCC. Either way I imagine these targets are almost exclusively used together with a gcc of some shape or form, so leaving it as `Some(8)` is fine in that case.

(Although amending a comment to that effect would be nice)

Comment (chrisnc) [reply compiler/rustc_target/src/spec/armv8r_none_eabihf.rs:27]: Happily. `gcc` has done this for arm-none-eabi going back to 2004-ish, and changed the  the `arm-linux` targets use `int` enum starting in 2005. I was unable to find evidence that `clang` ever had default `short-enum` for any Arm targets, though it does use it for some other targets.

File: compiler/rustc_target/src/spec/armv8r_none_eabihf.rs
Comment (chrisnc) [reply compiler/rustc_target/src/spec/armv8r_none_eabihf.rs:20]: Your interpretation is correct. I'll try and re-word it so it's clearer. The upstream web docs are IMO unclear because it's ambiguous whether neon and DP are coupled, but the manual is clear that there are exactly two configurations.

File: compiler/rustc_target/src/spec/armv8r_none_eabihf.rs
Comment (chrisnc) [reply compiler/rustc_target/src/spec/armv8r_none_eabihf.rs]: Updated all the comments claiming this that I could find, and removed one instance of this option and its comment that was not necessary due to using `thumb_base`.

File: compiler/rustc_target/src/spec/armv8r_none_eabihf.rs
Comment (chrisnc) [reply compiler/rustc_target/src/spec/armv8r_none_eabihf.rs]: Updated the comment describing the FP variants. Let me know if it's acceptable.

File: src/doc/rustc/src/platform-support.md
Comment (chrisnc) [reply src/doc/rustc/src/platform-support.md:150]: Speaking of... who are the maintainers of the various arm-none targets?

**Comment by chrisnc** [comment]
Somehow I'm having trouble getting the new `armv8r-none-eabihf` target to build, after rebasing, whereas I could build the cross compiler for the new target when I first made this change back in April. The issue seems to be when building `compiler-rt`; the build is selecting `/usr/bin/cc` for the target C compiler instead of `arm-none-eabi-gcc` like it does when building for the other arm-none targets, and then some of the assembly functions don't build because they're being interpreted for my host architecture. I don't see any obvious changes to the other targets that make this work.

I'm running `./x build --stage 1 --target armv8r-none-eabihf`, and it's failing on:
```
 error occurred: Command "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-ffunction-sections" "-fdata-sections" "-fno-builtin" "-fvisibility=hidden" "-ffreestanding" "-fomit-frame-pointer" "-ffile-prefix-map=/Users/chrisnc/src/rust/src/llvm-project/compiler-rt=." "-DVISIBILITY_HIDDEN" "-o" "/Users/chrisnc/src/rust/build/aarch64-apple-darwin/stage1-std/armv8r-none-eabihf/release/build/compiler_builtins-8c64fd81c6061885/out/aeabi_cdcmp.o" "-c" "/Users/chrisnc/src/rust/src/llvm-project/compiler-rt/lib/builtins/arm/aeabi_cdcmp.S" with args "cc" did not execute successfully (status code exit status: 1).
```

On the same commit, `./x build --stage 1 --target armv7r-none-eabihf` works, though, and it uses `arm-none-eabi-gcc` to build the `compiler-rt` code. Any suggestions for where to look would be appreciated.

**Comment by chrisnc** [comment]
Moved the minor changes to other targets to a separate PR which should be easier to merge: https://github.com/rust-lang/rust/pull/113992.

File: src/doc/rustc/src/platform-support.md
Comment (wesleywiser) [reply src/doc/rustc/src/platform-support.md:150]: Unfortunately we don't have great records here as many of these targets predate this process. For now, I would just document that you are the maintainer of this specific target.

File: compiler/rustc_target/src/spec/armv8r_none_eabihf.rs
Comment (nagisa) [reply compiler/rustc_target/src/spec/armv8r_none_eabihf.rs]: Yeah, the new comment is quite understandable.

**Comment by nagisa** [comment]
Code wise this LGTM.

r=me pending the checklist and a target page in the book. AFAIU cc-rs support does not actually block this target.

**Comment by JohnCSimon** [comment]
@chrisnc 
ping from triage - can you post your status on this PR? There hasn't been an update in a few months. Thanks!


FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

**Comment by chrisnc** [comment]
@JohnCSimon ack. My next step was to provide documentation for the other arm-none targets so it's clearer where this fits in with the more comprehensive target policy, because it doesn't quite make sense without all of that in place already (armv8r is a minor change that shares some aspects of armv7r and thumbv8em). I will update once that is ready.

**Comment by chrisnc** [comment]
Platform docs PR here: https://github.com/rust-lang/rust/pull/119102

Once that lands I will pick up this one again and give it its own md file and nominate it as a tier 3 target with the requested information.

**Comment by wllenyj** [comment]
@chrisnc 
Thanks for your pr. I'm looking forward to this merged. Instead of using armv7r/thumbv8m.

**Comment by chrisnc** [comment]
> A tier 3 target must have a designated developer or developers (the "target maintainers") on record to be CCed when issues arise regarding the target. (The mechanism to track and CC such developers may evolve over time.)

I've designated myself in the platform support document.

> Targets must use naming consistent with any existing targets; for instance, a target for the same CPU or OS as an existing Rust target should use the same name for that CPU or OS. Targets should normally use the same names and naming conventions as used elsewhere in the broader ecosystem beyond Rust (such as in other toolchains), unless they have a very good reason to diverge. Changing the name of a target can be highly disruptive, especially once the target reaches a higher tier, so getting the name right is important even for a tier 3 target.

This target follows the naming pattern of other arm-none targets and matches how other toolchains refer to it.

> Target names should not introduce undue confusion or ambiguity unless absolutely necessary to maintain ecosystem compatibility. For example, if the name of the target makes people extremely likely to form incorrect beliefs about what it targets, the name should be changed or augmented to disambiguate it.

Not applicable here. armv8r unambiguously refers to the 32-bit Arm architecture, version 8, R-profile.

> If possible, use only letters, numbers, dashes and underscores for the name. Periods (.) are known to cause issues in Cargo.

Check.

> Tier 3 targets may have unusual requirements to build or use, but must not create legal issues or impose onerous legal terms for the Rust project or for Rust developers or users.

Check.

> The target must not introduce license incompatibilities.

Check.

> Anything added to the Rust repository must be under the standard Rust license (MIT OR Apache-2.0).

Check.

> The target must not cause the Rust tools or libraries built for any other host (even when supporting cross-compilation to the target) to depend on any new dependency less permissive than the Rust licensing policy. This applies whether the dependency is a Rust crate that would require adding new license exceptions (as specified by the tidy tool in the rust-lang/rust repository), or whether the dependency is a native library or binary. In other words, the introduction of the target must not cause a user installing or running a version of Rust or the Rust tools to be subject to any new license requirements.

Check.

> Compiling, linking, and emitting functional binaries, libraries, or other code for the target (whether hosted on the target itself or cross-compiling from another target) must not depend on proprietary (non-FOSS) libraries. Host tools built for the target itself may depend on the ordinary runtime libraries supplied by the platform and commonly used by other applications built for the target, but those libraries must not be required for code generation for the target; cross-compilation to the target must not require such libraries at all. For instance, rustc built for the target may depend on a common proprietary C runtime library or console output library, but must not depend on a proprietary code generation library or code optimization library. Rust's license permits such combinations, but the Rust project has no interest in maintaining such combinations within the scope of Rust itself, even at tier 3.
"onerous" here is an intentionally subjective term. At a minimum, "onerous" legal/licensing terms include but are not limited to: non-disclosure requirements, non-compete requirements, contributor license agreements (CLAs) or equivalent, "non-commercial"/"research-only"/etc terms, requirements conditional on the employer or employment of any particular Rust developers, revocable terms, any requirements that create liability for the Rust project or its developers or users, or any requirements that adversely affect the livelihood or prospects of the Rust project or its developers or users.

Check.

> Neither this policy nor any decisions made regarding targets shall create any binding agreement or estoppel by any party. If any member of an approving Rust team serves as one of the maintainers of a target, or has any legal or employment requirement (explicit or implicit) that might affect their decisions regarding a target, they must recuse themselves from any approval decisions regarding the target's tier status, though they may otherwise participate in discussions.
This requirement does not prevent part or all of this policy from being cited in an explicit contract or work agreement (e.g. to implement or maintain support for a target). This requirement exists to ensure that a developer or team responsible for reviewing and approving a target does not face any legal threats or obligations that would prevent them from freely exercising their judgment in such approval, even if such judgment involves subjective matters or goes beyond the letter of these requirements.

Check.

> Tier 3 targets should attempt to implement as much of the standard libraries as possible and appropriate (core for most targets, alloc for targets that can support dynamic memory allocation, std for targets with an operating system or equivalent layer of system-provided functionality), but may leave some code unimplemented (either unavailable or stubbed out as appropriate), whether because the target makes it impossible to implement or challenging to implement. The authors of pull requests are not obligated to avoid calling any portions of the standard library on the basis of a tier 3 target not implementing those portions.

Par with other `arm-none` targets in this respect; `core`-only.

> The target must provide documentation for the Rust community explaining how to build for the target, using cross-compilation if possible. If the target supports running binaries, or running tests (even if they do not pass), the documentation must explain how to run such binaries or tests for the target, using emulation if possible or dedicated hardware if necessary.

Captured by arm-none platform support documentation and the additional platform support documentation in this PR.

> Tier 3 targets must not impose burden on the authors of pull requests, or other developers in the community, to maintain the target. In particular, do not post comments (automated or manual) on a PR that derail or suggest a block on the PR based on a tier 3 target. Do not send automated messages or notifications (via any medium, including via @) to a PR author or others involved with a PR regarding a tier 3 target, unless they have opted into such messages.
Backlinks such as those generated by the issue/PR tracker when linking to an issue or PR are not considered a violation of this policy, within reason. However, such messages (even on a separate repository) must not generate notifications to anyone involved with a PR who has not requested such notifications.

Check.

> Patches adding or updating tier 3 targets must not break any existing tier 2 or tier 1 target, and must not knowingly break another tier 3 target without approval of either the compiler team or the maintainers of the other tier 3 target.
In particular, this may come up when working on closely related targets, such as variations of the same architecture with different features. Avoid introducing unconditional uses of features that another variation of the target may not have; use conditional compilation or runtime detection, as appropriate, to let each target run code supported by that target.
Tier 3 targets must be able to produce assembly using at least one of rustc's supported backends from any host target.

This target is closely related to other arm-none targets, but does not interfere with them; the changes are strictly additive.

File: src/doc/rustc/src/platform-support.md
Comment (wesleywiser) [inline src/doc/rustc/src/platform-support.md:283]: nit: just keeping the list alphabetized 

```suggestion
[`armv7k-apple-watchos`](platform-support/apple-watchos.md) | ✓ |  | ARMv7-A Apple WatchOS
`armv7s-apple-ios` | ✓ |  | ARMv7-A Apple-A6 Apple iOS
[`armv8r-none-eabihf`](platform-support/armv8r-none-eabihf.md) | * |  | Bare ARMv8-R, hardfloat
```

**Review by wesleywiser — APPROVED** [review]

File: src/doc/rustc/src/platform-support.md
Comment (chrisnc) [reply src/doc/rustc/src/platform-support.md]: Whoops. Thanks.

**Comment by wesleywiser** [comment]
Thanks @chrisnc! 

@bors r+ rollup

**Comment by wllenyj** [comment]
Why can't the armv8r-none-eabihf target be installed by rustup?
It will note 
```
 $ rustup target add armv8r-none-eabihf
error: component 'rust-std' for target 'armv8r-none-eabihf' is unavailable for download for channel 'stable'

$ rustc -vV
rustc 1.79.0 (129f3b996 2024-06-10)
binary: rustc
commit-hash: 129f3b9964af4d4a709d1383930ade12dfe7c081
commit-date: 2024-06-10
host: aarch64-apple-darwin
release: 1.79.0
LLVM version: 18.1.7
```

**Comment by chrisnc** [comment]
@wllenyj this is because it's currently a Tier 3 target, so pre-built `core` is not available.
https://doc.rust-lang.org/nightly/rustc/target-tier-policy.html

You can build it yourself using:
https://doc.rust-lang.org/cargo/reference/unstable.html#build-std

**Comment by wllenyj** [comment]
> @wllenyj this is because it's currently a Tier 3 target, so pre-built `core` is not available. https://doc.rust-lang.org/nightly/rustc/target-tier-policy.html
> 
> You can build it yourself using: https://doc.rust-lang.org/cargo/reference/unstable.html#build-std

oh, I got, thank you very much.

**Pull request merged**
